crates/starknet-devnet/tests/test_simulate_transactions.rs

```diff
@@ -479,4 +493,568 @@ mod simulation_tests {
+    }
+
+    #[tokio::test]
+    async fn simulate_invoke_v3_with_fee_just_below_estimated_should_return_a_trace_of_reverted_transaction()
+     {
+        let devnet = BackgroundDevnet::spawn().await.expect("Could not start Devnet");
+        let (sierra_artifact, casm_hash) =
+            get_flattened_sierra_contract_and_casm_hash(CAIRO_1_PANICKING_CONTRACT_SIERRA_PATH);
+
+        let (signer, account_address) = devnet.get_first_predeployed_account().await;
+
+        let mut account = SingleOwnerAccount::new(
+            &devnet.json_rpc_client,
+            signer,
+            account_address,
+            constants::CHAIN_ID,
+            ExecutionEncoding::New,
+        );
+        account.set_block_id(BlockId::Tag(BlockTag::Latest));
+
+        let declare_result =
+            account.declare_v3(Arc::new(sierra_artifact.clone()), casm_hash).send().await.unwrap();
+
+        let salt = Felt::from_hex_unchecked("0x123");
+        let execution = account.execute_v3(vec![Call {
+            to: UDC_CONTRACT_ADDRESS,
+            selector: get_selector_from_name("deployContract").unwrap(),
+            calldata: vec![
+                declare_result.class_hash,
+                salt,
+                Felt::ZERO, // is_unique
+                Felt::ZERO, // constructor data length
+            ],
+        }]);
+        let fee_estimate = execution.estimate_fee().await.unwrap();
+
+        let (gas_units, gas_price) = get_gas_units_and_gas_price(fee_estimate);
+
+        let SimulatedTransaction { transaction_trace, .. } =
+            execution.gas(gas_units - 1).gas_price(gas_price).simulate(false, false).await.unwrap();
+
+        match transaction_trace {
+            TransactionTrace::Invoke(InvokeTransactionTrace {
+                execute_invocation: ExecuteInvocation::Reverted(reverted_invocation),
+                ..
+            }) => assert_contains(&reverted_invocation.revert_reason, "Insufficient"),
+            other => panic!("Unexpected trace {other:?}"),
+        }
+    }
+
+    #[tokio::test]
+    async fn simulate_invoke_v3_with_failing_execution_should_return_a_trace_of_reverted_transaction()
+     {
+        let devnet = BackgroundDevnet::spawn().await.expect("Could not start Devnet");
+        let (sierra_artifact, casm_hash) =
+            get_flattened_sierra_contract_and_casm_hash(CAIRO_1_PANICKING_CONTRACT_SIERRA_PATH);
+
+        let (signer, account_address) = devnet.get_first_predeployed_account().await;
+
+        let mut account = SingleOwnerAccount::new(
+            &devnet.json_rpc_client,
+            signer,
+            account_address,
+            constants::CHAIN_ID,
+            ExecutionEncoding::New,
+        );
+        account.set_block_id(BlockId::Tag(BlockTag::Latest));
+
+        let declare_result =
+            account.declare_v3(Arc::new(sierra_artifact.clone()), casm_hash).send().await.unwrap();
+
+        let salt = Felt::from_hex_unchecked("0x123");
+
+        account
```

Comment:
There is a function in commons/utils.rs that abstracts declaring and deploying. Currently it does so relying on v1, but that shouldn't matter in this and similar cases in this file, because the declaring and deploying don't need to be v3 if we're later just testing a panic behavior in an invoke v3.